### PR TITLE
brave.exe -> chrome.exe mimic.

### PR DESCRIPTION
**Pull request by boocmp**
Resolves https://github.com/brave/brave-browser/issues/29113

image

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Enable brave-module-filename-patch flag in the brave://flags
Open brave://gpu, find `Executable path` section, should point to `chrome.exe` instead of `brave.exe`

### CODE REVIEW

File: sandbox/win/src/module_file_name_interception.cc [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+namespace sandbox {
+
+SANDBOX_INTERCEPT DWORD WINAPI
+TargetGetModuleFileNameA(GetModuleFileNameAFunction orig,
+                         HMODULE hModule,
+                         LPSTR lpFilename,
+                         DWORD nSize) {
+  auto result = orig(hModule, lpFilename, nSize);
+  if (hModule == nullptr && result != 0) {
```
Comment (goodov) [inline sandbox/win/src/module_file_name_interception.cc:77]: `hModule == nullptr || hModule == GetModuleHandle(nullptr)` ?

File: sandbox/win/src/module_file_name_interception.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+#define TargetGetModuleFileNameExA TargetK32GetModuleFileNameExA
+#define TargetGetModuleFileNameExW TargetK32GetModuleFileNameExW
+#endif
+
+namespace sandbox {
+
+using GetModuleFileNameAFunction = decltype(&::GetModuleFileNameA);
+using GetModuleFileNameWFunction = decltype(&::GetModuleFileNameW);
+using GetModuleFileNameExAFunction = decltype(&::GetModuleFileNameExA);
+using GetModuleFileNameExWFunction = decltype(&::GetModuleFileNameExW);
```
Comment (goodov) [inline sandbox/win/src/module_file_name_interception.h:25]: there's also `GetModuleBaseName`, but MSDN recommends `GetModuleFileName` to get the current process name.

**Review by goodov — APPROVED** [review]

File: sandbox/win/src/module_file_name_interception.h [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+#include "sandbox/win/src/sandbox_types.h"
+
+#if (PSAPI_VERSION > 1)
+#define TargetGetModuleFileNameExA TargetK32GetModuleFileNameExA
+#define TargetGetModuleFileNameExW TargetK32GetModuleFileNameExW
+#endif
+
+namespace sandbox {
+
+BASE_DECLARE_FEATURE(kModuleFileNamePatch);
```
Comment (iefremov) [inline sandbox/win/src/module_file_name_interception.h:22]: comment

File: chromium_src/components/viz/service/gl/gpu_service_impl.h [hunk 4]
```diff
@@ -0,0 +1,24 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_CHROMIUM_SRC_COMPONENTS_VIZ_SERVICE_GL_GPU_SERVICE_IMPL_H_
+#define BRAVE_CHROMIUM_SRC_COMPONENTS_VIZ_SERVICE_GL_GPU_SERVICE_IMPL_H_
+
+#define InitializeWithHost                                             \
+  InitializeWithHost_ChromiumImpl(                                     \
```
Comment (iefremov) [inline chromium_src/components/viz/service/gl/gpu_service_impl.h:10]: pls add a comment reflecting for what the override is used

File: sandbox/win/src/module_file_name_interception.cc [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+  }
+
+  constexpr DWORD kBraveLen = BraveToChrome<CharT>::kBrave.length();
+  constexpr DWORD kChromeLen = BraveToChrome<CharT>::kChrome.length();
+  static_assert(kBraveLen <= kChromeLen);
+  constexpr DWORD kLenDiff = kChromeLen - kBraveLen;
+
+  const size_t brave_pos = length - kBraveLen;
+  BraveToChrome<CharT>::ReplaceAt(filename + brave_pos, size - brave_pos);
+  if (size < length + kLenDiff) {
```
Comment (iefremov) [inline sandbox/win/src/module_file_name_interception.cc:60]: i guess it's better to do this check first, then replace? Also you can return after `setlasterror`?

File: chromium_src/services/viz/privileged/mojom/gl/gpu_host.mojom [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+module viz.mojom;
+
+[BraveExtend]
+interface GpuHost {
+  [EnableIf=is_win]
+  DidGetExecutablePath(string path);
```
Comment (iefremov) [inline chromium_src/services/viz/privileged/mojom/gl/gpu_host.mojom:11]: comment

**Review by iefremov — APPROVED** [review]

**Comment by iefremov** [comment]
it was marked as qa/yes, so provide test steps for QAs

File: sandbox/win/src/module_file_name_interception.cc (hunk 5 above)
Comment (boocmp) [reply sandbox/win/src/module_file_name_interception.cc:60]: No, you're wrong, read msdn

Comment (iefremov) [reply sandbox/win/src/module_file_name_interception.cc:58]: write a comment then

File: sandbox/win/src/module_file_name_interception.cc [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+template <typename CharT>
+struct BraveToChrome;
+
+template <>
+struct BraveToChrome<char> {
+  static constexpr const base::StringPiece kBrave = "brave.exe";
+  static constexpr const base::StringPiece kChrome = "chrome.exe";
+
+  static void ReplaceAt(char* dest, size_t dest_size) {
+    ::strncpy_s(dest, dest_size, kChrome.data(), kChrome.length());
```
Comment (iefremov) [inline sandbox/win/src/module_file_name_interception.cc:29]: It fills dest with zeroes if `dest_size <= count`

**Pull request merged**